# Chapter1 

# 第一节：数据加载及初步观察

## 1.1 载入数据

### 1.1.1 任务一：导入numpy和pandas

In [1]:
import numpy as np
import pandas as pd

### 1.1.2 任务二：载入数据

In [3]:
df=pd.read_csv('/Users/linjiaxi/Desktop/学习/hands on data analysis/chapter 1/titanic_data/train.csv')
df.head(3)

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S

### 1.1.3 任务三：每1000行为一个数据模块，逐块读取

In [6]:
#利用逐块读取
chunker=pd.read_csv('/Users/linjiaxi/Desktop/学习/hands on data analysis/chapter 1/titanic_data/train.csv',chunksize=1000)

for chunk in chunker:
    #在这里进行对每个块的操作
    #可以对块进行数据分析、处理等
    #每个快的数据存储在chunk变量中
    
    print(chunk.head())

   PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S  


逐块读取（Chunking）是指将大型数据文件分割成较小的块（或批次），然后逐个处理这些块<br>

逐块处理的优点：<br>
1.内存效率：当处理大型数据文件时，一次性将整个文件加载到内存可能会导致内存不足的问题。通过逐块读取，你可以将数据分成较小的块，并逐个处理，从而减少对内存的要求。<br>

2.增量处理：逐块读取允许你逐个处理数据块，并对每个块执行特定的操作。这对于实时或增量处理数据非常有用，因为你可以在处理一个块后立即获取结果，而无需等待整个数据集加载和处理完成。<br>

3.迭代处理：通过逐块读取，你可以使用循环来处理每个块。这意味着你可以在处理一个块后进行一些操作，然后再处理下一个块，以此类推。这种迭代方式非常灵活，允许你在处理数据时逐步构建结果或执行复杂的数据转换。<br>

### 1.1.4 任务四：将表头改成中文，索引改为乘客ID

PassengerId => 乘客ID<br>
Survived => 是否幸存<br>
Pclass => 乘客等级(1/2/3等舱位)<br>
Name => 乘客姓名<br>
Sex => 性别<br>
Age => 年龄<br>
SibSp => 堂兄弟/妹个数<br>
Parch => 父母与小孩个数<br>
Ticket => 船票信息<br>
Fare => 票价<br>
Cabin => 客舱<br>
Embarked => 登船港口<br>

#### 方法一：

In [7]:
df=pd.read_csv('/Users/linjiaxi/Desktop/学习/hands on data analysis/chapter 1/titanic_data/train.csv',
               names=['乘客ID','是否幸存','仓位等级','姓名','性别','年龄','兄弟姐妹个数','父母子女个数','船票信息','票价','客舱'])
df.head()

乘客ID    是否幸存  \
PassengerId  Survived  Pclass   
1                   0       3   
2                   1       1   
3                   1       3   
4                   1       1   

                                                          仓位等级      姓名   性别  \
PassengerId                                               Name     Sex  Age   
1                                      Braund, Mr. Owen Harris    male   22   
2            Cumings, Mrs. John Bradley (Florence Briggs Th...  female   38   
3                                       Heikkinen, Miss. Laina  female   26   
4                 Futrelle, Mrs. Jacques Heath (Lily May Peel)  female   35   

                年龄 兄弟姐妹个数            父母子女个数     船票信息     票价        客舱  
PassengerId  SibSp  Parch            Ticket     Fare  Cabin  Embarked  
1                1      0         A/5 21171     7.25    NaN         S  
2                1      0          PC 17599  71.2833    C85         C  
3                0      0  STON/O2. 3101282    7.925    NaN         S  
4                1      0            113803     53.1   C123         S

#### 方法二：

In [8]:
df=pd.read_csv('/Users/linjiaxi/Desktop/学习/hands on data analysis/chapter 1/titanic_data/train.csv')
new_columns=['乘客ID','是否幸存','仓位等级','姓名','性别','年龄','兄弟姐妹个数','父母子女个数','船票信息','票价','客舱']
df.rename(columns=dict(zip(df.columns,new_columns)),inplace=True)
df.head(3)

乘客ID  是否幸存  仓位等级                                                 姓名  \
0     1     0     3                            Braund, Mr. Owen Harris   
1     2     1     1  Cumings, Mrs. John Bradley (Florence Briggs Th...   
2     3     1     3                             Heikkinen, Miss. Laina   

       性别    年龄  兄弟姐妹个数  父母子女个数              船票信息       票价   客舱 Embarked  
0    male  22.0       1       0         A/5 21171   7.2500  NaN        S  
1  female  38.0       1       0          PC 17599  71.2833  C85        C  
2  female  26.0       0       0  STON/O2. 3101282   7.9250  NaN        S

## 1.2 初步观察

### 1.2.1 任务一：查看数据的基本信息

In [9]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   乘客ID      891 non-null    int64  
 1   是否幸存      891 non-null    int64  
 2   仓位等级      891 non-null    int64  
 3   姓名        891 non-null    object 
 4   性别        891 non-null    object 
 5   年龄        714 non-null    float64
 6   兄弟姐妹个数    891 non-null    int64  
 7   父母子女个数    891 non-null    int64  
 8   船票信息      891 non-null    object 
 9   票价        891 non-null    float64
 10  客舱        204 non-null    object 
 11  Embarked  889 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB


### 1.2.2 任务二：观察表格前10行的数据和后15行的数据

In [10]:
df.head(10)

乘客ID  是否幸存  仓位等级                                                 姓名  \
0     1     0     3                            Braund, Mr. Owen Harris   
1     2     1     1  Cumings, Mrs. John Bradley (Florence Briggs Th...   
2     3     1     3                             Heikkinen, Miss. Laina   
3     4     1     1       Futrelle, Mrs. Jacques Heath (Lily May Peel)   
4     5     0     3                           Allen, Mr. William Henry   
5     6     0     3                                   Moran, Mr. James   
6     7     0     1                            McCarthy, Mr. Timothy J   
7     8     0     3                     Palsson, Master. Gosta Leonard   
8     9     1     3  Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)   
9    10     1     2                Nasser, Mrs. Nicholas (Adele Achem)   

       性别    年龄  兄弟姐妹个数  父母子女个数              船票信息       票价    客舱 Embarked  
0    male  22.0       1       0         A/5 21171   7.2500   NaN        S  
1  female  38.0       1       0          PC 17599  71.2833   C85        C  
2  female  26.0       0       0  STON/O2. 3101282   7.9250   NaN        S  
3  female  35.0       1       0            113803  53.1000  C123        S  
4    male  35.0       0       0            373450   8.0500   NaN        S  
5    male   NaN       0       0            330877   8.4583   NaN        Q  
6    male  54.0       0       0             17463  51.8625   E46        S  
7    male   2.0       3       1            349909  21.0750   NaN        S  
8  female  27.0       0       2            347742  11.1333   NaN        S  
9  female  14.0       1       0            237736  30.0708   NaN        C

In [11]:
df.tail(15)

乘客ID  是否幸存  仓位等级                                             姓名      性别  \
876   877     0     3                  Gustafsson, Mr. Alfred Ossian    male   
877   878     0     3                           Petroff, Mr. Nedelio    male   
878   879     0     3                             Laleff, Mr. Kristo    male   
879   880     1     1  Potter, Mrs. Thomas Jr (Lily Alexenia Wilson)  female   
880   881     1     2   Shelley, Mrs. William (Imanita Parrish Hall)  female   
881   882     0     3                             Markun, Mr. Johann    male   
882   883     0     3                   Dahlberg, Miss. Gerda Ulrika  female   
883   884     0     2                  Banfield, Mr. Frederick James    male   
884   885     0     3                         Sutehall, Mr. Henry Jr    male   
885   886     0     3           Rice, Mrs. William (Margaret Norton)  female   
886   887     0     2                          Montvila, Rev. Juozas    male   
887   888     1     1                   Graham, Miss. Margaret Edith  female   
888   889     0     3       Johnston, Miss. Catherine Helen "Carrie"  female   
889   890     1     1                          Behr, Mr. Karl Howell    male   
890   891     0     3                            Dooley, Mr. Patrick    male   

       年龄  兄弟姐妹个数  父母子女个数              船票信息       票价    客舱 Embarked  
876  20.0       0       0              7534   9.8458   NaN        S  
877  19.0       0       0            349212   7.8958   NaN        S  
878   NaN       0       0            349217   7.8958   NaN        S  
879  56.0       0       1             11767  83.1583   C50        C  
880  25.0       0       1            230433  26.0000   NaN        S  
881  33.0       0       0            349257   7.8958   NaN        S  
882  22.0       0       0              7552  10.5167   NaN        S  
883  28.0       0       0  C.A./SOTON 34068  10.5000   NaN        S  
884  25.0       0       0   SOTON/OQ 392076   7.0500   NaN        S  
885  39.0       0       5            382652  29.1250   NaN        Q  
886  27.0       0       0            211536  13.0000   NaN        S  
887  19.0       0       0            112053  30.0000   B42        S  
888   NaN       1       2        W./C. 6607  23.4500   NaN        S  
889  26.0       0       0            111369  30.0000  C148        C  
890  32.0       0       0            370376   7.7500   NaN        Q

### 1.2.4 任务三：判断数据是否为空，为空的地方返回True，其余地方返回False

In [13]:
df.isnull().head()

乘客ID   是否幸存   仓位等级     姓名     性别     年龄  兄弟姐妹个数  父母子女个数   船票信息     票价  \
0  False  False  False  False  False  False   False   False  False  False   
1  False  False  False  False  False  False   False   False  False  False   
2  False  False  False  False  False  False   False   False  False  False   
3  False  False  False  False  False  False   False   False  False  False   
4  False  False  False  False  False  False   False   False  False  False   

      客舱  Embarked  
0   True     False  
1  False     False  
2   True     False  
3  False     False  
4   True     False

## 1.3 保存数据

### 1.3.1 任务一：将你加载并做出改变的数据，在工作目录下保存为一个新文件train_chinese.csv

In [14]:
df.to_csv('/Users/linjiaxi/Desktop/学习/hands on data analysis/chapter 1/titanic_data/train_chinese.csv',encoding='GBK')

# 第二节：Pandas基础

## 1.4 知道你的数据叫什么

### 1.4.1 任务一：pandas中有两个数据类型DateFrame和Series，通过查找简单了解他们。

In [15]:
import numpy as np
import pandas as pd

In [16]:
#Series
sdata = {'Ohio': 35000, 
         'Texas': 71000, 
         'Oregon': 16000,
         'Utah': 5000}
example_1 = pd.Series(sdata)
example_1

Ohio      35000
Texas     71000
Oregon    16000
Utah       5000
dtype: int64

In [17]:
#DataFrame
data = {'state': ['Ohio', 'Ohio', 'Ohio', 'Nevada', 'Nevada', 'Nevada'],
        'year': [2000, 2001, 2002, 2001, 2002, 2003],
        'pop': [1.5, 1.7, 3.6, 2.4, 2.9, 3.2]}
example_2 = pd.DataFrame(data)
example_2

state  year  pop
0    Ohio  2000  1.5
1    Ohio  2001  1.7
2    Ohio  2002  3.6
3  Nevada  2001  2.4
4  Nevada  2002  2.9
5  Nevada  2003  3.2

### 1.4.2 任务二：根据上节课的方法载入"train.csv"文件

In [18]:
df=pd.read_csv('/Users/linjiaxi/Desktop/学习/hands on data analysis/chapter 1/titanic_data/train.csv')
df.head(3)

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S

### 1.4.3 任务三：查看DataFrame数据的每列的名称

In [20]:
df.columns

Index(['PassengerId', 'Survived', 'Pclass', 'Name', 'Sex', 'Age', 'SibSp',
       'Parch', 'Ticket', 'Fare', 'Cabin', 'Embarked'],
      dtype='object')

### 1.4.4任务四：查看"Cabin"这列的所有值 

In [21]:
#方法一：
df['Cabin'].head(3)

0    NaN
1    C85
2    NaN
Name: Cabin, dtype: object

In [22]:
#方法二：
df.Cabin.head(3)

0    NaN
1    C85
2    NaN
Name: Cabin, dtype: object

### 1.4.5 任务五：加载文件"test_1.csv"，然后对比"train.csv"，看看有哪些多出的列，然后将多出的列删除

In [24]:
test=pd.read_csv('/Users/linjiaxi/Desktop/学习/hands on data analysis/chapter 1/titanic_data/test.csv')
test_1=test.copy()
test_1.head(3)
#test没有survived列

PassengerId  Pclass                              Name     Sex   Age  SibSp  \
0          892       3                  Kelly, Mr. James    male  34.5      0   
1          893       3  Wilkes, Mrs. James (Ellen Needs)  female  47.0      1   
2          894       2         Myles, Mr. Thomas Francis    male  62.0      0   

   Parch  Ticket    Fare Cabin Embarked  
0      0  330911  7.8292   NaN        Q  
1      0  363272  7.0000   NaN        S  
2      0  240276  9.6875   NaN        Q

### 1.4.6 任务六： 将['PassengerId','Name','Age','Ticket']这几个列元素隐藏，只观察其他几个列元素

In [25]:
df.drop(['PassengerId','Name','Age','Ticket'],axis=1).head(3)

Survived  Pclass     Sex  SibSp  Parch     Fare Cabin Embarked
0         0       3    male      1      0   7.2500   NaN        S
1         1       1  female      1      0  71.2833   C85        C
2         1       3  female      0      0   7.9250   NaN        S

In [26]:
df.head(3)

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S

备注:<br>
我们并没有完全从df中删除'PassengerId','Name','Age','Ticket'列，只是隐藏了

## 1.5 筛选的逻辑

### 1.5.1 任务一： 我们以"Age"为筛选条件，显示年龄在10岁以下的乘客信息。

In [27]:
df[df['Age']<10].head(3)
#中间的df['Age']<10为筛选条件

PassengerId  Survived  Pclass                             Name     Sex  \
7             8         0       3   Palsson, Master. Gosta Leonard    male   
10           11         1       3  Sandstrom, Miss. Marguerite Rut  female   
16           17         0       3             Rice, Master. Eugene    male   

    Age  SibSp  Parch   Ticket    Fare Cabin Embarked  
7   2.0      3      1   349909  21.075   NaN        S  
10  4.0      1      1  PP 9549  16.700    G6        S  
16  2.0      4      1   382652  29.125   NaN        Q

### 1.5.2 任务二： 以"Age"为条件，将年龄在10岁以上和50岁以下的乘客信息显示出来，并将这个数据命名为midage

In [28]:
#条件交集的使用
midage = df[(df["Age"]>10)&(df["Age"]<50)]
midage.head(3)

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S

### 1.5.3 任务三：将midage的数据中第100行的"Pclass"和"Sex"的数据显示出来

In [30]:
#并未重新设置索引
midage.loc[[100],['Pclass','Sex']]

Pclass     Sex
100       3  female

#### reset_index函数的作用<br>
reset_index() 是 Pandas 中的一个函数，用于重置数据帧（DataFrame）的索引。<br>
它的作用是将当前的索引重新设置为默认的整数索引，并将原来的索引作为一个新的列添加到数据帧中。<br>
默认情况下，数据帧的索引是整数序列（0、1、2、...），如果在处理数据时对索引进行了修改或者使用了某些操作导致索引不再是连续的整数序列，那么可以使用 reset_index() 函数将索引重置为默认的整数序列

In [31]:
#重新设置索引
midage = midage.reset_index(drop=True)
#利用drop参数丢弃原来的索引，重新设置索引
midage.head(3)
midage.loc[[100],['Pclass','Sex']]

Pclass   Sex
100       2  male

#### loc函数的作用<br>
通过行标签和列标签来访问数据帧（DataFrame）中的特定行和列，或者对它们进行切片和筛选

### 1.5.4 任务四：使用loc方法将midage的数据中第100，105，108行的"Pclass"，"Name"和"Sex"的数据显示出来

In [32]:
midage.loc[[100,105,108],['Pclass','Name','Sex']] 

Pclass                               Name   Sex
100       2  Byles, Rev. Thomas Roussel Davids  male
105       3           Cribb, Mr. John Hatfield  male
108       3                    Calic, Mr. Jovo  male

### 1.5.5 任务五：使用iloc方法将midage的数据中第100，105，108行的"Pclass"，"Name"和"Sex"的数据显示出来

In [33]:
midage.iloc[[100,105,108],[2,3,4]]

Pclass                               Name   Sex
100       2  Byles, Rev. Thomas Roussel Davids  male
105       3           Cribb, Mr. John Hatfield  male
108       3                    Calic, Mr. Jovo  male

#### iloc与loc的异同

- **索引方式**：
   - `iloc` 使用基于位置的索引，即通过整数位置来定位行和列。
   - `loc` 使用基于标签的索引，即通过行标签和列标签来定位行和列。

- **使用方式**：
   - `iloc` 的使用方式类似于Python中的列表索引，它通过指定整数位置来选择数据。
   - `loc` 的使用方式更接近于字典的键值对，它通过指定行标签和列标签来选择数据。

- **切片操作**：
   - `iloc` 使用的是半开区间（左闭右开）的切片操作。
   - `loc` 使用的是闭区间的切片操作。

- **性能**：
   - `iloc` 在使用整数位置索引时通常会比 `loc` 快。
   - `loc` 在使用标签索引时可能稍慢，因为它需要进行额外的标签匹配操作。

# 第三节：探索性数据分析

In [34]:
import numpy as np
import pandas as pd

In [39]:
text=pd.read_csv('/Users/linjiaxi/Desktop/学习/hands on data analysis/chapter 1/titanic_data/train.csv')
new_columns=['乘客ID','是否幸存','仓位等级','姓名','性别','年龄','兄弟姐妹个数','父母子女个数','船票信息','票价','客舱']
text.rename(columns=dict(zip(text.columns,new_columns)),inplace=True)
text.head(3)

乘客ID  是否幸存  仓位等级                                                 姓名  \
0     1     0     3                            Braund, Mr. Owen Harris   
1     2     1     1  Cumings, Mrs. John Bradley (Florence Briggs Th...   
2     3     1     3                             Heikkinen, Miss. Laina   

       性别    年龄  兄弟姐妹个数  父母子女个数              船票信息       票价   客舱 Embarked  
0    male  22.0       1       0         A/5 21171   7.2500  NaN        S  
1  female  38.0       1       0          PC 17599  71.2833  C85        C  
2  female  26.0       0       0  STON/O2. 3101282   7.9250  NaN        S

## 1.6 了解你的数据吗？

### 1.6.1 任务一：利用Pandas对示例数据进行排序，要求升序

In [41]:
#自行构建的数据
frame=pd.DataFrame(np.arange(8).reshape((2,4)),
#reshape相当于设置frame为2行4列的dataframe
index=['2','1'],
columns=['d','a','b','c']
#columns则设置列名称)
frame

d  a  b  c
2  0  1  2  3
1  4  5  6  7

pd.DataFrame() ：创建一个DataFrame对象<br>

np.arange(8).reshape((2, 4)) : 生成一个二维数组（2*4）,第一列：0，1，2，3 第二列：4，5，6，7<br>

index=['2','1'] ：DataFrame 对象的索引列<br>

columns=['d', 'a', 'b', 'c'] ：DataFrame 对象的索引行<br>

In [43]:
frame.sort_values(by='c',ascending=False)

d  a  b  c
1  4  5  6  7
2  0  1  2  3

sort_values这个函数中by参数指向要排列的列<br>
ascending参数指向排序的方式（升序还是降序）

In [44]:
# 让行索引升序排序
frame.sort_index()

d  a  b  c
1  4  5  6  7
2  0  1  2  3

In [45]:
# 让列索引升序排序
frame.sort_index(axis=1)

a  b  c  d
2  1  2  3  0
1  5  6  7  4

In [46]:
# 让列索引降序排序
frame.sort_index(axis=1, ascending=False)

d  c  b  a
2  0  3  2  1
1  4  7  6  5

In [47]:
# 让任选两列数据同时降序排序
frame.sort_values(by=['a', 'c'], ascending=False)

d  a  b  c
1  4  5  6  7
2  0  1  2  3

### 1.6.2 任务二：对泰坦尼克号数据（trian.csv）按票价和年龄两列进行综合排序（降序排列），从数据中你能发现什么

In [49]:
text.sort_values(by=['票价', '年龄'], ascending=False)

乘客ID  是否幸存  仓位等级                                  姓名      性别    年龄  \
679   680     1     1  Cardeza, Mr. Thomas Drake Martinez    male  36.0   
258   259     1     1                    Ward, Miss. Anna  female  35.0   
737   738     1     1              Lesurer, Mr. Gustave J    male  35.0   
438   439     0     1                   Fortune, Mr. Mark    male  64.0   
341   342     1     1      Fortune, Miss. Alice Elizabeth  female  24.0   
..    ...   ...   ...                                 ...     ...   ...   
481   482     0     2    Frost, Mr. Anthony Wood "Archie"    male   NaN   
633   634     0     1       Parr, Mr. William Henry Marsh    male   NaN   
674   675     0     2          Watson, Mr. Ennis Hastings    male   NaN   
732   733     0     2                Knight, Mr. Robert J    male   NaN   
815   816     0     1                    Fry, Mr. Richard    male   NaN   

     兄弟姐妹个数  父母子女个数      船票信息        票价           客舱 Embarked  
679       0       1  PC 17755  512.3292  B51 B53 B55        C  
258       0       0  PC 17755  512.3292          NaN        C  
737       0       0  PC 17755  512.3292         B101        C  
438       1       4     19950  263.0000  C23 C25 C27        S  
341       3       2     19950  263.0000  C23 C25 C27        S  
..      ...     ...       ...       ...          ...      ...  
481       0       0    239854    0.0000          NaN        S  
633       0       0    112052    0.0000          NaN        S  
674       0       0    239856    0.0000          NaN        S  
732       0       0    239855    0.0000          NaN        S  
815       0       0    112058    0.0000         B102        S  

[891 rows x 12 columns]

分析:<br>
我们知道票价越高的舱位应该越好，那么我们可以看到票价高的幸存率明显高于票价低的，这是票价与幸存之间的关系<br>
**那么年龄与幸存之间的关系呢？**

### 1.6.3 任务三：利用Pandas进行算术计算，计算两个DataFrame数据相加结果

In [50]:
#自行建立例子
frame1_a = pd.DataFrame(np.arange(9.).reshape(3, 3),
                     columns=['a', 'b', 'c'],
                     index=['one', 'two', 'three'])
frame1_b = pd.DataFrame(np.arange(12.).reshape(4, 3),
                     columns=['a', 'e', 'c'],
                     index=['first', 'one', 'two', 'second'])
frame1_a

a    b    c
one    0.0  1.0  2.0
two    3.0  4.0  5.0
three  6.0  7.0  8.0

In [51]:
frame1_b

a     e     c
first   0.0   1.0   2.0
one     3.0   4.0   5.0
two     6.0   7.0   8.0
second  9.0  10.0  11.0

In [52]:
#将frame_a和frame_b进行相加
frame1_a + frame1_b

a   b     c   e
first   NaN NaN   NaN NaN
one     3.0 NaN   7.0 NaN
second  NaN NaN   NaN NaN
three   NaN NaN   NaN NaN
two     9.0 NaN  13.0 NaN

两个DataFrame相加后，会返回一个新的DataFrame，对应的行和列的值会相加，没有对应的会变成空值NaN

### 1.6.4 任务四：通过泰坦尼克号数据如何计算出在船上最大的家族有多少人？

In [53]:
max(text['兄弟姐妹个数'] + text['父母子女个数'])

10

### 1.6.5 任务五：学会使用Pandas describe()函数查看数据基本统计信息

In [54]:
#建立例子
frame2 = pd.DataFrame([[1.4, np.nan], 
                       [7.1, -4.5],
                       [np.nan, np.nan], 
                       [0.75, -1.3]
                      ], index=['a', 'b', 'c', 'd'], columns=['one', 'two'])
frame2

one  two
a  1.40  NaN
b  7.10 -4.5
c   NaN  NaN
d  0.75 -1.3

In [56]:
# 调用 describe 函数，观察frame2的数据基本信息

frame2.describe()

one       two
count  3.000000  2.000000
mean   3.083333 -2.900000
std    3.493685  2.262742
min    0.750000 -4.500000
25%    1.075000 -3.700000
50%    1.400000 -2.900000
75%    4.250000 -2.100000
max    7.100000 -1.300000

count : 样本数据大小<br>
mean : 样本数据的平均值<br>
std : 样本数据的标准差<br>
min : 样本数据的最小值<br>
25% : 样本数据25%的时候的值<br>
50% : 样本数据50%的时候的值<br>
75% : 样本数据75%的时候的值<br>
max : 样本数据的最大值


### 1.6.6 任务六：分别看看泰坦尼克号数据集中 票价、父母子女 这列数据的基本统计数据，你能发现什么？

In [57]:
text['票价'].describe()

count    891.000000
mean      32.204208
std       49.693429
min        0.000000
25%        7.910400
50%       14.454200
75%       31.000000
max      512.329200
Name: 票价, dtype: float64

一共有891个票价数据<br>
平均值约为：32.20<br>
标准差约为49.69，说明票价波动特别大<br>
25%的人的票价是低于7.91的，50%的人的票价低于14.45，75%的人的票价低于31.00<br>
票价最大值约为512.33，最小值为0。 

In [58]:
text['父母子女个数'].describe()

count    891.000000
mean       0.381594
std        0.806057
min        0.000000
25%        0.000000
50%        0.000000
75%        0.000000
max        6.000000
Name: 父母子女个数, dtype: float64